File: lib/rating/rating.dart
import "package:angular2/angular2.dart";
import 'dart:html';
import 'package:node_shims/js.dart';

// TODO: templateUrl
@Component (selector: "rating[ng-model]",
    inputs: const [
      "max", "readonly", "titles", "stateOn", "stateOff", "ratingStates"],
    outputs: const [ "onHover", "onLeave"],
    host: const { "(keydown)" : "onKeydown(\$event)"},
    template: '''
    <span (mouseleave)="reset()" (keydown)="onKeydown(\$event)" tabindex="0" role="slider" aria-valuemin="0" [attr.aria-valuemax]="range.length" [attr.aria-valuenow]="value">
      <template ng-for #r [ng-for-of]="range" #index="index">
        <span class="sr-only">({{ index < value ? '*' : ' ' }})</span>
        <i (mouseenter)="enter(index + 1)" (click)="rate(index + 1)" class="glyphicon" [ng-class]="index < value ? r['stateOn'] : r['stateOff']" [title]="r['title']" ></i>
      </template>
    </span>
  ''', directives: const [NgClass, NgFor])
class Rating extends DefaultValueAccessor implements OnInit {
  Rating(this.cd, Renderer renderer, ElementRef elementRef)
      : super (renderer, elementRef) {
    cd.valueAccessor = this;
  }

  NgModel cd;

  num max;

  List range;

  num value;

  num preValue;

  List<String> titles;

  String stateOn;

  String stateOff;

  bool readonly;

  List ratingStates;

  EventEmitter onHover = new EventEmitter ();

  EventEmitter onLeave = new EventEmitter ();

  onInit() {
    max ??= 5;
    readonly = readonly == true;
    stateOn ??= "glyphicon-star";
    stateOff ??= "glyphicon-star-empty";
    titles = titles != null && titles.length > 0  ? titles : ["one", "two", "three", "four", "five"];
    ratingStates ??= [];
    range = _buildTemplateObjects();
  }

  // model -> view
  writeValue(num _value) {
    _value ??= 0;
    if (_value != 0) {
      value = _value.round();
      preValue = _value;
      return;
    }
    preValue = _value;
    value = _value;

  }

  _buildTemplateObjects() {
    var count = or(ratingStates.length, max) ;
    var result = [];
    for (var i = 0; i < count; i++) {
      result.add({
        "index" : i,
        "stateOn" : stateOn,
        "stateOff" : stateOff,
        "title" : titles.length > i ? titles[i] : i + 1,
      }..addAll(ratingStates.length > i ? ratingStates[i] : {}));
    }
    return result;
  }

  rate(num value) {
    if (!readonly && value >= 0 && value <= range.length) {
      writeValue(value);
      cd.viewToModelUpdate(value);
    }
  }

  enter(num _value) {
    if (!readonly) {
      value = _value;
      onHover.add(_value);
    }
  }

  reset() {
    value = preValue;
    onLeave.add(value);
  }

  onKeydown(KeyboardEvent event) {
    if (![37, 38, 39, 40].contains(event.which)) {
      return;
    }
    event.preventDefault();
    event.stopPropagation();
    var sign = event.which == 38 || event.which == 39
        ? 1
        : -1;
    rate(value + sign);
  }
}

File: lib/rating/rating.ts
/// <reference path="../../tsd.d.ts" />

import {
  Component, View,
  OnInit, EventEmitter,
  NgClass, DefaultValueAccessor, NgFor,
  NgModel, Self, Renderer, ElementRef
} from 'angular2/angular2';

// TODO: templateUrl
@Component({
  selector: 'rating[ng-model]',
  properties: [
    'max', 'readonly', 'titles',
    'stateOn', 'stateOff',
    'ratingStates'
  ],
  events: ['onHover', 'onLeave'],
  host: {
    '(keydown)': 'onKeydown($event)'
  }
})
@View({
  template: `
    <span (mouseleave)="reset()" (keydown)="onKeydown($event)" tabindex="0" role="slider" aria-valuemin="0" [attr.aria-valuemax]="range.length" [attr.aria-valuenow]="value">
      <template ngFor #r [ngForOf]="range" #index="index">
        <span class="sr-only">({{ index < value ? '*' : ' ' }})</span>
        <i (mouseenter)="enter(index + 1)" (click)="rate(index + 1)" class="glyphicon" [ngClass]="index < value ? r.stateOn : r.stateOff" [title]="r.title" ></i>
      </template>
    </span>
  `,
  directives: [NgClass, NgFor]
})
export class Rating extends DefaultValueAccessor implements OnInit {
  private max:number;
  private range:Array<any>;
  private value:number;
  private preValue:number;
  private titles:Array<string>;
  private stateOn:string;
  private stateOff:string;
  private readonly:boolean;
  private ratingStates:Array<{stateOn:string, stateOff:string}>;

  private onHover:EventEmitter = new EventEmitter();
  private onLeave:EventEmitter = new EventEmitter();

  constructor(@Self() cd:NgModel, renderer:Renderer, elementRef:ElementRef) {
    super(cd, renderer, elementRef);
  }

  ngOnInit() {
    this.max = typeof this.max !== 'undefined' ? this.max : 5;
    this.readonly = this.readonly === true;
    this.stateOn = typeof this.stateOn !== 'undefined' ? this.stateOn : 'glyphicon-star';
    this.stateOff = typeof this.stateOff !== 'undefined' ? this.stateOff : 'glyphicon-star-empty';
    this.titles = typeof this.titles !== 'undefined' && this.titles.length > 0 ? this.titles : ['one', 'two', 'three', 'four', 'five'];
    this.range = this.buildTemplateObjects(this.ratingStates, this.max);
  }

  // model -> view
  writeValue(value:number) {
    if (value % 1 !== value) {
      this.value = Math.round(value);
      this.preValue = value;
      return;
    }

    this.preValue = value;
    this.value = value;
  }

  private buildTemplateObjects(ratingStates:Array<any>, max:number) {
    ratingStates = ratingStates || [];
    let count = ratingStates.length || max;
    let result = [];
    for (let i = 0; i < count; i++) {
      result.push(Object.assign({
        index: i,
        stateOn: this.stateOn,
        stateOff: this.stateOff,
        title: this.titles[i] || i + 1
      }, ratingStates[i] || {}));
    }
    return result;
  }

  private rate(value:number) {
    if (!this.readonly && value >= 0 && value <= this.range.length) {
      this.writeValue(value);
      this.cd.viewToModelUpdate(value);
    }
  }

  private enter(value:number) {
    if (!this.readonly) {
      this.value = value;
      this.onHover.next(value);
    }
  }

  private reset() {
    this.value = this.preValue;
    this.onLeave.next(this.value);
  }

  private onKeydown(event:KeyboardEvent) {
    if ([37, 38, 39, 40].indexOf(event.which) === -1) {
      return;
    }

    event.preventDefault();
    event.stopPropagation();
    let sign = event.which === 38 || event.which === 39 ? 1 : -1;
    this.rate(this.value + sign);
  }
}
